pkg/auth/providers/ldap/ldap_client.go

```diff
@@ -166,9 +167,62 @@ func (p *ldapProvider) getPrincipalsFromSearchResult(result *ldapv2.SearchResult
 		}
 	}
 
+	allowedPrincipalIDs := config.AllowedPrincipalIDs
+	for _, allowedPid := range allowedPrincipalIDs {
+		parts := strings.SplitN(allowedPid, ":", 2)
+		if len(parts) != 2 {
+			return userPrincipal, groupPrincipals, errors.Errorf("invalid id %v", allowedPid)
+		}
+		scope := parts[0]
+		entityType := strings.Split(scope, "_")[1]
+		entityDN := strings.TrimPrefix(parts[1], "//")
+
+		searchDomain := config.UserSearchBase
+		if config.GroupSearchBase != "" {
+			searchDomain = config.GroupSearchBase
+		}
+
+		if strings.EqualFold("group", entityType) {
+			err := p.findNestedGroups(searchDomain, entityDN, userDN, groupScope, config, lConn, userPrincipal)
+			if err != nil {
+				return userPrincipal, groupPrincipals, err
+			}
+		}
+	}
 	return userPrincipal, groupPrincipals, nil
 }
 
+func (p *ldapProvider) findNestedGroups(groupSearchDomain string, groupDN string, userDN string, groupScope string, config *v3.LdapConfig, lConn *ldapv2.Conn, userPrincipal v3.Principal) error {
+	searchGroup := ldapv2.NewSearchRequest(groupSearchDomain,
+		ldapv2.ScopeWholeSubtree, ldapv2.NeverDerefAliases, 0, 0, false,
+		fmt.Sprintf("(&(%v=%v)(objectClass=%v))", config.GroupDNAttribute, groupDN, config.GroupObjectClass),
+		p.getGroupSearchAttributes(config), nil)
+	resultGroup, err := lConn.Search(searchGroup)
+	if err != nil {
+		return err
+	}
+
+	if len(resultGroup.Entries) > 0 {
+		for _, entry := range resultGroup.Entries {
+			entryAttr := entry.Attributes
+			for _, attr := range entryAttr {
+				if attr.Name == "member" {
+					for _, currentDN := range attr.Values {
+						if currentDN == userDN {
+							//add this user to allowedPrincipalID list
+							config.AllowedPrincipalIDs = append(config.AllowedPrincipalIDs, userPrincipal.ObjectMeta.Name)
+						} else {
```

Comment:
And, instead of iterating over the config.AllowedPrincipalID list and expanding it (to check if any of its members is a group having this user as a member); can we traverse the tree from the user group up until you reach the parent group?

So consider this ldap setup:

* G1
    * | ---- G2
         * | ------G3
             * | ---- User U1

and config.AllowedPrincipalIDs = {G1}

So when the User U1 logs in, you have the info that  U1 is memberOf G3.
Then find if G3 is a 'member' of any groupObject - you will get G2, add it to groupPrincipals of U1.
Then check if G2 is a 'member' of any groupObject - you will get G1, add it to groupPrincipals of U1.
Then check if G1 is a 'member' of any groupObject, so on until you reach the root base,
In the end you get   groupPrincipals of U1 = {G3, G2, G1}